cloud/pkg/admissioncontroller/admission_test.go

```diff
@@ -0,0 +1,681 @@
+
+	assert.Equal(t, 1, len(mutateWebhook2.Webhooks))
+	assert.Equal(t, MutatingNodeUpgradeWebhookName, mutateWebhook2.Webhooks[0].Name)
+}
+
+func TestGetRuleEndpoint(t *testing.T) {
+	ac := &AdmissionController{
+		CrdClient: versioned_fake.NewSimpleClientset(),
+	}
+
+	name := "test-endpoint"
+
+	ruleEndpoint := &v1.RuleEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: defaultNamespace,
+		},
+	}
+	_, err := ac.CrdClient.RulesV1().RuleEndpoints(defaultNamespace).Create(context.TODO(), ruleEndpoint, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	result, err := ac.getRuleEndpoint(defaultNamespace, name)
+	assert.NoError(t, err)
+	assert.Equal(t, name, result.Name)
+
+	result, err = ac.getRuleEndpoint(defaultNamespace, "nonexistent-endpoint")
+	assert.Error(t, err)
+	assert.Nil(t, result)
+}
+
+func TestListRule(t *testing.T) {
+	ac := &AdmissionController{
+		CrdClient: versioned_fake.NewSimpleClientset(),
+	}
+
+	rule1 := &v1.Rule{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rule1",
+			Namespace: defaultNamespace,
+		},
+	}
+	rule2 := &v1.Rule{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rule2",
+			Namespace: defaultNamespace,
+		},
+	}
+
+	_, err := ac.CrdClient.RulesV1().Rules(defaultNamespace).Create(
+		context.TODO(), rule1, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	_, err = ac.CrdClient.RulesV1().Rules(defaultNamespace).Create(
+		context.TODO(), rule2, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	result, err := ac.listRule(defaultNamespace)
+	assert.NoError(t, err)
+	assert.Len(t, result, 2)
+}
+
+func TestListRuleEndpoint(t *testing.T) {
+	ac := &AdmissionController{
+		CrdClient: versioned_fake.NewSimpleClientset(),
+	}
+
+	endpoint1 := &v1.RuleEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "endpoint1",
+			Namespace: defaultNamespace,
+		},
+	}
+	endpoint2 := &v1.RuleEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "endpoint2",
+			Namespace: defaultNamespace,
+		},
+	}
+
+	_, err := ac.CrdClient.RulesV1().RuleEndpoints(defaultNamespace).Create(
+		context.TODO(), endpoint1, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	_, err = ac.CrdClient.RulesV1().RuleEndpoints(defaultNamespace).Create(
+		context.TODO(), endpoint2, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	result, err := ac.listRuleEndpoint(defaultNamespace)
+	assert.NoError(t, err)
+	assert.Len(t, result, 2)
+}
+
+func TestRunWithFakeClients(t *testing.T) {
+	opt := &options.AdmissionOptions{
+		Kubeconfig:                "/non/existent/path/to/kubeconfig",
+		CaCertFile:                "/non/existent/path/to/cacert",
+		CertFile:                  "/non/existent/path/to/cert",
+		KeyFile:                   "/non/existent/path/to/key",
+		AdmissionServiceName:      "test-service",
+		AdmissionServiceNamespace: defaultNamespace,
+	}
+
+	err := Run(opt)
+	assert.Error(t, err)
+
+	opt.Kubeconfig = ""
+	opt.CaCertFile = ""
+	opt.CertFile = ""
+	opt.KeyFile = ""
```

Comment:
```suggestion
	opt.Kubeconfig, opt.CaCertFile ..... = "",""......
```
Is this more concise?